EIPS/eip-2348.md

```diff
@@ -0,0 +1,228 @@
+opcode location. Clients either need to do the analysis and store the values or re-evaluate the
+contract on each execution. Stronger deployment validation will allow clients to presume jump calls
+are valid in certain circumstances.
+
+A tertiary motivation is to prepare the way for easily JITable contracts. While the current EVM can
+be JIT compiled there are certain analyses that need to be performed to prevent or accommodate some
+pathological or uncompilable cases from being compiled. With stricter rules these cases can be
+detected at deploy time and rejected allowing EVM clients to make better assumptions about the
+contract being compiled.
+
+## Specification
+
+There are three interlocking portions specified in this EIP and two portions from other active EIPs
+included in this validation. [EIP-1702] (Generalized Account Versioning Scheme) and [EIP-2327]
+(`BEGINDATA` opcode) are specified in their published locations. The portions specified in this EIP
+are a versioning header (similar to what was in [EIP-1707]), invalid opcode validation (similar to
+[EIP-1712]), and static jump analysis.
+
+### EVM Account Versioning
+
+Starting at `BLOCKNUM` (TBD) `EIP-1702` will be activated, `LATEST_VERSION` will be set to `1`, and
+all new and updated accounts will have the account version `1`. The validation phase will apply the
+rules described in the Version Header, `BEGINDATA`, Invalid Opcode Validation, and Static Jump
+Validations sections.
+
+These EIP sections only apply to contracts stored or in the process of being stored in in accounts
+with version `1`. This EIP never applies to contracts stored or in the process of being stored in
+accounts at version `0`. Future EIPs may increase the set of contract versions this EIP applies to.
+
+### Version Header
+
+For contracts with the first byte from `0x01` to `0xff`, or whose total length is less than 4 bytes,
+the contract is treated exactly as through it had been deployed to an account with version `0`. For
+these contracts none of the other subsections in this EIP apply.
+
+When deploying a contract if a contract starts with `0x00` and has a length 4 or later the first
+four bytes form a version header. If a version header is not recognized by the EVM the contract
+deployment transaction fails with out-of-gas.
+
+For this EIP, only header version `1` (contracts starting with the byte stream `0x00` `0x00` `0x00`
+`0x01`) is defined. Future EIPs may expand on the valid set of headers. This version indicates that
+next three validations are applied to the content of the contract, keeping all other semantics of
+the current "version 0" EVM contracts, including the same gas schedule.
+
+For purposes of EVM Program Counter calculations the first byte after the version header is location
+`0`. The contract header is not part of the accessible contract data. There is no mechanism within
+the EVM to retrieve the header values.
```

Comment:
>  There is no mechanism within the EVM to retrieve the header values.

`CODECOPY`?

Reply:
Good catch, but how does this reconcile with the start location?  Do we keep PC=0 at zero and specify the EVM should skip the first 4 bytes of such a contract?  Or do we let it not match?